echo_server.c

```diff
@@ -49,65 +57,109 @@ static int send_request(struct socket *sock, unsigned char *buf, size_t size)
 
     printk(MODULE_NAME ": start send request.\n");
 
+    /* FIXME: what if somehow length of message is zero? */
     length = kernel_sendmsg(sock, &msg, &vec, 1, strlen(buf) - 1);
 
     printk(MODULE_NAME ": send request = %s\n", buf);
 
     return length;
 }
 
-static int echo_server_worker(void *arg)
+static void echo_server_worker(struct work_struct *work)
 {
-    struct socket *sock;
+    struct fastecho *worker =
+        container_of(work, struct fastecho, fastecho_work);
     unsigned char *buf;
     int res;
 
-    sock = (struct socket *) arg;
-    allow_signal(SIGKILL);
-    allow_signal(SIGTERM);
-
     buf = kmalloc(BUF_SIZE, GFP_KERNEL);
     if (!buf) {
         printk(KERN_ERR MODULE_NAME ": kmalloc error....\n");
-        return -1;
+        return;
     }
 
-    while (!kthread_should_stop()) {
-        res = get_request(sock, buf, BUF_SIZE - 1);
+    /*
+     * FIXME: first echo message may contains unexpected characters
```

Comment:
Should we clear the buffer every time before calling `kernel_recvmsg`?

Reply:
Yes, it would be great to do this, but this is left for intention (for homework).